pkg/types/installconfig.go

```diff
@@ -120,6 +120,11 @@ type InstallConfig struct {
 	// +optional
 	FIPS bool `json:"fips,omitempty"`
 
+	// InstallType provides signal to the cluster that the installation will not follow the default installer workflow.
+	//
+	// +optional
+	InstallType string `json:"installType,omitempty"`
```

Comment:
I feel like there's a more descriptive name to be found, and it should be an enum at least... seems like it will inevitably become mirror of PlatformType to some degree but with different lifecycle implications.

Is a raw enum going to be enough? (It wasn't for platform type.) What's the relationship between this type and and platform types? Which combinations are supported, and what are the clear semantics and cluster implications for each type and combination of type/platform? etc.

Would a vendor annotation on the installconfig make any sense as an alternative if we aren't able to more carefully specify these things right now?

Reply:
(see: https://github.com/openshift/enhancements/pull/480#discussion_r493765988)

Reply:
> Is a raw enum going to be enough? (It wasn't for platform type.) What's the relationship between this type and and platform types? Which combinations are supported, and what are the clear semantics and cluster implications for each type and combination of type/platform? etc.

I agree with you in general but the idea is that the logic and relationships that each operator would conclude should be based on the value. We could certainly make this more strict and only define explicitly supported types. Then extended the list as necessary.

> Would a vendor annotation on the installconfig make any sense as an alternative if we aren't able to more carefully specify these things right now?

If no agreement can be made I am open to any reasonable signal.

Reply:
> (see: openshift/enhancements#480 (comment))

I agree that tying the install type to a platform doesn't make sense. The idea is to completely platform agnostic as the install type is not related to the platform it is performed on the platform. That was my intention in making the key at root level.

Reply:
> I feel like there's a more descriptive name to be found, and it should be an enum at least

If you have a better name I am open to change.

Reply:
I'm most interested in clarity of semantics. Possible values, what they mean on their own, what they mean with respect to platform types.

For example, based on what I understand so far, here's a try:

```go

// AssistedInstaller indicates the platform will assume an installation
// topology consisting of exactly 3 nodes which will be the only available
// nodes to complete cluster bootstrapping.
//
// The platform may relax certain high availability constraints to achieve
// bootstrapping under the constrained topology.
//
// This install type is only supported when the platform type is BareMetal.
type AssistedInstaller InstallType = "AssistedInstaller"
```

Reply:
Maybe it needs to be even more specific in terms of the machine pools. Isn't the requirement for there to be a single `master` pool with 3 replicas and an _empty_ `worker` pool? This should be unambiguous.

Reply:
why? we let the user choose how many workers he adds - among all the discovered hardware, we auto-select 3 masters and all the rest become workers

Reply:
Good point. Is the absence of the bootstrap node the substantive difference? Again, we should figure out how to define it

Reply:
Worker pools are not assumed but unlike compact clusters workers are possible. @romfreiman keep me honest here.

Reply:
The assisted installer API is a wrapper for openshift-install, so it seems a little strange to define an install-config interface for this, given that end-users interacting with that config currently won't have any use for this option.

I wonder could this be expressed by some kind of annotation on the machine objects themselves?

Then the assisted flow could write that annotation during the process of generating the manifests/install-config (using the installer generated data as an input), and add the necessary annotation?

That said, I can imagine in future we may want to adopt a similar approach for IPI baremetal (which currently requires a host to run a bootstrap VM alongside the installer), but in that case I imagine we'd either just use the assisted flow to run the bootstrap services, or some subset of the same workflow resulting in a similarly modified bootstrap environment?